services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/ChecksumIntegrationTest.java

```diff
@@ -0,0 +1,133 @@
+import software.amazon.awssdk.services.s3.model.Tag;
+import software.amazon.awssdk.services.s3.model.Tagging;
+import software.amazon.awssdk.testutils.RandomTempFile;
+import software.amazon.awssdk.testutils.service.AwsTestBase;
+
+public class ChecksumIntegrationTest extends S3IntegrationTestBase {
+    private static final String TEST_BUCKET = temporaryBucketName(ChecksumIntegrationTest.class);
+    private static final String TEST_KEY = "10mib_file.dat";
+    private static final int OBJ_SIZE = 1024 * 1024;
+
+    private static RandomTempFile testFile;
+    private static S3CrtAsyncClient s3Crt;
+
+    @BeforeAll
+    public static void setup() throws Exception {
+        S3IntegrationTestBase.setUp();
+        S3IntegrationTestBase.createBucket(TEST_BUCKET);
+
+        testFile = new RandomTempFile(TEST_KEY, OBJ_SIZE);
+
+        s3Crt = S3CrtAsyncClient.builder()
+                                .credentialsProvider(AwsTestBase.CREDENTIALS_PROVIDER_CHAIN)
+                                .region(S3IntegrationTestBase.DEFAULT_REGION)
+                                .build();
+    }
+
+    @AfterAll
+    public static void teardown() throws IOException {
+        S3IntegrationTestBase.deleteBucketAndAllContents(TEST_BUCKET);
+        Files.delete(testFile.toPath());
+        s3Crt.close();
+        CrtResource.waitForNoResources();
+    }
+
+    @Test
+    public void noChecksumCustomization_crc32ShouldBeUsed() {
+        AsyncRequestBody body = AsyncRequestBody.fromFile(testFile.toPath());
+        PutObjectResponse putObjectResponse =
+            s3Crt.putObject(r -> r.bucket(TEST_BUCKET).key(TEST_KEY), body).join();
+        String putObjectChecksum = putObjectResponse.checksumCRC32();
+        assertThat(putObjectChecksum).isNotNull();
+
+        ResponseBytes<GetObjectResponse> getObjectResponseResponseBytes =
+            s3Crt.getObject(r -> r.bucket(TEST_BUCKET).key(TEST_KEY), AsyncResponseTransformer.toBytes()).join();
+        String getObjectChecksum = getObjectResponseResponseBytes.response().checksumCRC32();
+        assertThat(getObjectChecksum).isNotNull();
+        assertThat(putObjectChecksum).isEqualTo(getObjectChecksum);
+    }
+
+    @Test
+    public void checksumProvidedInRequest_shouldTakePrecendence() {
+        AsyncRequestBody body = AsyncRequestBody.fromFile(testFile.toPath());
+        PutObjectResponse putObjectResponse =
+            s3Crt.putObject(r -> r.bucket(TEST_BUCKET).key(TEST_KEY).checksumAlgorithm(ChecksumAlgorithm.SHA1), body).join();
+        String putObjectChecksum = putObjectResponse.checksumSHA1();
+        assertThat(putObjectChecksum).isNotNull();
+
+        ResponseBytes<GetObjectResponse> getObjectResponseResponseBytes =
+            s3Crt.getObject(r -> r.bucket(TEST_BUCKET).key(TEST_KEY), AsyncResponseTransformer.toBytes()).join();
+        String getObjectChecksum = getObjectResponseResponseBytes.response().checksumSHA1();
```

Comment:
It's not strictly part of the PR, but I'm wondering about how customers who do a get without having done a put can know which checksum to get from the response. Do you think it's assumed that either they have a checksum from somewhere and then they know which one to use, or if they don't they won't require it?

Reply:
In most cases, customers do not need to worry about validating checksum because the SDK knows the algorithm from the response headers and validates it automatically. The SDK also maintains a list of response algorithms that should be used if there are multiple checksums available in the response headers.

I think users can do a headObject call to know if flexible checksum is enabled or not.

Reply:
ok!